dandiapi/api/management/commands/correct_metadata.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import TYPE_CHECKING, Any
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+from django.db import transaction
+import djclick as click
+
+from dandiapi.api.models import Version
+from dandiapi.api.tasks import write_manifest_files
+
+
+@click.command(
+    help='Correct corrupted metadata. If `--dandiset` and `--dandiset-version` are provided, '
+    'apply the correction to the specified Dandiset version. Otherwise, apply the correction to '
+    'all Dandiset versions.'
+)
+@click.option(
+    '--dandiset',
+    'dandiset',
+    help='The Dandiset. (This must be provided if `--dandiset-version` is provided.)',
+)
+@click.option(
+    '--dandiset-version',
+    'dandiset_version',
+    help='The version of the Dandiset. (This must be provided if `--dandiset` is provided.)',
+)
+def correct_metadata(*, dandiset: str | None, dandiset_version: str | None):
+    # Func to use to correct the metadata (update as needed)
+    correct_func: Callable[[dict], dict | None] = correct_affiliation_corruption
+
+    # Ensure both options are provided together (or both omitted)
+    if (dandiset is None) != (dandiset_version is None):
+        raise click.UsageError(
+            'Both `--dandiset` and `--dandiset-version` must be provided together.'
+        )
+
+    vers = (
+        Version.objects.all()
+        if dandiset is None
+        else (Version.objects.get(dandiset=dandiset, version=dandiset_version),)
+    )
+
+    # List of changes to be applied to the database represented as tuples of the form
+    # version to be changed and the new metadata
+    changes: list[tuple[Version, dict]] = []
+    for ver in vers:
+        try:
+            meta_corrected = correct_func(ver.metadata)
+        except Exception as e:
+            click.echo(
+                f'{ver.dandiset.identifier}/{ver.version}: Failed to correct metadata with '
+                f'`{correct_func.__name__}()`'
+            )
+            click.echo(e)
+            raise click.Abort from e
+
+        if meta_corrected is not None:
+            changes.append((ver, meta_corrected))
+
+    if changes:
+        # Apply the changes to the database atomically
+        with transaction.atomic():
+            for ver, meta_corrected in changes:
+                ver.metadata = meta_corrected
+                ver.save()  # TODO: should I use `save()` in this case? It has some custom logic.
+
+        for ver, _ in changes:
+            write_manifest_files.delay(ver.id)  # TODO: Please check if this is needed
```

Comment:
What happens if any of the file writes fail or time out? https://github.com/dandi/dandi-archive/blob/7728016558ab7ee77ee06534d461b119a200cbc4/dandiapi/api/tasks/__init__.py#L58-L62

Maybe we should at least keep the AsyncResults and monitor that they go as planned?

Reply:
I actually don't know if I need to call `write_manifest_files.delay(ver.id)` at all. I was trying to mimic https://github.com/dandi/dandi-archive/blob/7728016558ab7ee77ee06534d461b119a200cbc4/dandiapi/api/management/commands/migrate_published_version_metadata.py#L52, but https://github.com/dandi/dandi-archive/blob/master/dandiapi/api/management/commands/migrate_version_metadata.py didn't call `write_manifest_files()` at all. 

> Maybe we should at least keep the AsyncResults and monitor that they go as planned?

I think the answer is related to how important that statement gets executed successfully. In the example above, its execution status is not checked. If a success execution is important, I guess I can't interact with the DB the current way.

@dandi/dandiarchive